### ISSUE TITLE
twisted/internet/process.py leaks file descriptors when os.fork() raises

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2305]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2305|
|Type|defect|
|Created|2006-12-19 13:05:19Z|
|Branch|https://github.com/twisted/twisted/tree/pipe-leak-2305-3|

os.pipe() is called before os.fork()

No error handling is done around os.fork() or os.pipe().

Any file descriptors created for pipes before the os.fork() will be leaked indefinitely if os.fork()/pipe() raises an exception.

<details><summary>Searchable metadata</summary>

```
trac-id__2305 2305
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_pipe_leak_2305_3 branches/pipe-leak-2305-3
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1166533519000000 1166533519000000
changetime__1212483028000000 1212483028000000
version__None None
owner__ 
cc__eric.newton cc__therve cc__itamarst
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

[#2335](https://github.com/twisted/twisted/issues/2335), a duplicate of this ticket, includes a patch.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

Updated to note os.pipe() has no error handling either.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve removed owner|
|-|-|

Ready to review in pipe-leak-2305.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @therve|
|-|-|

This is going to conflict pretty badly with [[[#2341](https://github.com/twisted/twisted/issues/2341)](https://github.com/twisted/twisted/issues/2341)](https://github.com/twisted/twisted/issues/2341).  It also only fixes the problem for spawnProcess(usePTY=False), not spawnProcess(usePTY=True) (and the duplication in code/bugs currently involved in implementing those two variations is one reason #2341 is a good idea).

Recommend holding off on this until #2341 is merged, and then fixing it once.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

Suggested improvement for this code: instead of re-raising errors, pass them as Failures to processEnded, so user can handle them.

As far as finishing [[#2341](https://github.com/twisted/twisted/issues/2341)](https://github.com/twisted/twisted/issues/2341): at least some of it has been moved out ([[#2419](https://github.com/twisted/twisted/issues/2419)](https://github.com/twisted/twisted/issues/2419), awaiting review) and some of it still needs moving out ([[#2420](https://github.com/twisted/twisted/issues/2420)](https://github.com/twisted/twisted/issues/2420)). The remaining code is probably still useful until we do the full rewrite, but will need some work (removing stuff that is now in other bugs, writing some tests, validating it all works by someone who knows low-level OS details).

My suggested order: #2419, #2420, #2341, then this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

This is perhaps a duplicate of [#10979](https://github.com/twisted/twisted/issues/10979)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [23625]) Branching to 'pipe-leak-2305-3'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

This is ready to review. Another side effect I solved was with UID management: the original settings were not restored in case of failure.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @therve|
|-|-|

  * test_process.py
    * `test_mockForkErrorClosePTY` has a typo - _C{use**r**PTY}_
    * `test_mockErrorInPipe` has a typo - _the create**s** pipes_
    * I think `set` makes sense for comparing things like `self.mockos.closed` and `[-1, -4, -6, -2, -3, -5]` (in `test_mockForErrorClosesFDs) - the order doesn't actually matter here, right?  Likewise for the other comparions of lists of file descriptors which appear in several of the other test methods.
    * in the docstring for `test_mockErrorInForkRestoreUID`, I'd write ''a UID change'' rather than ''an UID change''
  * process.py
    * I think the UID/GID should be restored before `gc.enable()` is called.  The order for the non-exceptional case seems wrong to me as well.  `gc.enable()` might call arbitrary application code, it'd be bad to do that with the wrong credentials I bet.
    * Maybe try/finally could be used to avoid duplicating the cleanup code?
    * I'm not sure there's any way the `os.close` calls at line 544, 545, 825, and 826 could raise an exception, but I'm not sure they certainly can't either.  Worst case, they could fail with something like `MemoryError`.  Hmm.  I was going to suggest giving them exception handling and being careful that the original exception gets raised at the end, but maybe we should just wait and see if there's an actual reason to do this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

One of the tests fails for me:

```
[FAIL]: twisted.test.test_process.MockProcessTestCase.test_mockForkErrorClosePTY

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/test/test_process.py", line 1577, in test_mockForkErrorClosePTY
    self.assertEqual(self.mockos.closed, [4, 9])
twisted.trial.unittest.FailTest: not equal:
a = [4, 7]
b = [4, 9]

```

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_13"></a>|@therve commented|
|-|-|

(In [23820]) Process review comments.

Refs [#2305](https://github.com/twisted/twisted/issues/2305)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve removed owner|
|-|-|

I've fixed every points. I didn't used try/finally because it was a bit complicated to figure out out to fix the tests in client side.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set owner to @therve|
|-|-|

Looks good.  No worries about the try/finally thing.  I noticed another issue, I think.  If `childFDs` is specified, then file descriptors from it will end up in the `fdmap` local and then get closed if `_fork` fails.  I think it would be better if they were left open. What do you think?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve commented|
|-|-|

(In [23831]) Fix the problem with user-specified fd

Refs [#2305](https://github.com/twisted/twisted/issues/2305)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_17"></a>|@therve removed owner|
|-|-|

That seems like a good suggestion :) Solved now I think.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun set owner to @therve|
|-|-|

In the comment in `test_mockForkErrorGivenFDs`, _creates_ should be _create_.  That's it, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_19"></a>|@therve set status to `closed`|
|-|-|

(In [23846]) Merge pipe-leak-2305-3

Author: therve
Reviewer: exarkun
Fixes [#2305](https://github.com/twisted/twisted/issues/2305)
Fixes [#10979](https://github.com/twisted/twisted/issues/10979)

Handle errors raised by os.pipe and os.fork in spawnProcess, to make sure that
opened file descriptors are closed and that UID settings are restored.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
